File: Swift/YourLawyer/YourLawyer/Model/Post.swift
//
//  Post.swift
//  YourLawyer
//
//  Created by Alan Ulises on 15/01/25.
//

import Foundation

func performPostRequest(
	urlString: String,
	headers: [String: String],
	bodyParameters: [String: Any],
	completion: @escaping (Result<Data, Error>) -> Void
) {
	// Configurar la URL
	guard let url = URL(string: urlString) else {
		print("URL no válida")
		return
	}
	
	// Configurar el cuerpo de la solicitud
	var bodyComponents = URLComponents()
	bodyComponents.queryItems = bodyParameters.map { key, value in
		URLQueryItem(name: key, value: "\(value)")
	}
	
	// Configurar la solicitud
	var request = URLRequest(url: url)
	request.httpMethod = "POST"
	request.httpBody = bodyComponents.query?.data(using: .utf8)
	
	// Agregar encabezados
	for (key, value) in headers {
		request.setValue(value, forHTTPHeaderField: key)
	}
	
	// Crear la tarea de red
	let task = URLSession.shared.dataTask(with: request) { data, response, error in
		if let error = error {
			completion(.failure(error))
			return
		}
		
		guard let httpResponse = response as? HTTPURLResponse, (200...299).contains(httpResponse.statusCode) else {
			let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0
			let error = NSError(domain: "", code: statusCode, userInfo: [NSLocalizedDescriptionKey: "Solicitud fallida con código de estado: \(statusCode)"])
			completion(.failure(error))
			return
		}
		
		if let data = data {
			completion(.success(data))
		}
	}
	
	task.resume()
}

func createEphemeralKey(for customerID: String, completion: @escaping (Result<Data, Error>) -> Void) {
	let url = ephemeralURL
	let headers = [
		"Content-Type": "application/x-www-form-urlencoded",
		"Stripe-Version": "2022-11-15",
		"Authorization": "Basic \(stripeKeySecret.data(using: .utf8)!.base64EncodedString())"
	]
	let body = [
		"customer": customerID
	]
	
	performPostRequest(urlString: url, headers: headers, bodyParameters: body, completion: completion)
}

func createPaymentIntent(amount: String, currency: String, completion: @escaping (Result<Data, Error>) -> Void) {
	let url = clientSecretURL
	let headers = [
		"Content-Type": "application/x-www-form-urlencoded",
		"Authorization": "Basic \(stripeKeySecret.data(using: .utf8)!.base64EncodedString())"
	]
	let body = [
		"amount": amount,
		"currency": currency,
		"automatic_payment_methods[enabled]": "true"
	]
	
	performPostRequest(urlString: url, headers: headers, bodyParameters: body, completion: completion)
}

File: Kotlin/app/src/main/java/mx/com/yourlawyer/yourlawyer/controller/Post.kt
package mx.com.yourlawyer.yourlawyer.controller

import android.os.Build
import androidx.annotation.RequiresApi
import mx.com.yourlawyer.yourlawyer.utils.Constants.clientSecretURL
import mx.com.yourlawyer.yourlawyer.utils.Constants.ephemeralURL
import mx.com.yourlawyer.yourlawyer.utils.Constants.stripeKeySecret
import okhttp3.*
import okhttp3.MediaType.Companion.toMediaType
import java.io.IOException
import java.util.Base64

val JSON: MediaType = "application/json; charset=utf-8".toMediaType()
val client = OkHttpClient()

fun performPostRequest(
    urlString: String,
    headers: Map<String, String>,
    bodyParameters: Map<String, String>,
    completion: (Result<String>) -> Unit
) {
    val formBodyBuilder = FormBody.Builder()
    for ((key, value) in bodyParameters) {
        formBodyBuilder.add(key, value)
    }
    val formBody = formBodyBuilder.build()

    val requestBuilder = Request.Builder().url(urlString).post(formBody)
    for ((key, value) in headers) {
        requestBuilder.addHeader(key, value)
    }
    val request = requestBuilder.build()

    client.newCall(request).enqueue(object : Callback {
        override fun onFailure(call: Call, e: IOException) {
            completion(Result.failure(e))
        }

        override fun onResponse(call: Call, response: Response) {
            if (!response.isSuccessful) {
                val statusCode = response.code
                val error = IOException("Request failed with status code: $statusCode")
                completion(Result.failure(error))
                return
            }

            val responseData = response.body?.string() ?: ""
            completion(Result.success(responseData))
        }
    })
}

@RequiresApi(Build.VERSION_CODES.O)
fun createEphemeralKey(forCustomerID: String, completion: (Result<String>) -> Unit) {
    val headers = mapOf(
        "Content-Type" to "application/x-www-form-urlencoded",
        "Stripe-Version" to "2022-11-15",
        "Authorization" to "Basic ${Base64.getEncoder().encodeToString(stripeKeySecret.toByteArray(Charsets.UTF_8))}"
    )
    val body = mapOf("customer" to forCustomerID)

    performPostRequest(ephemeralURL, headers, body, completion)
}

@RequiresApi(Build.VERSION_CODES.O)
fun createPaymentIntent(amount: String, currency: String, completion: (Result<String>) -> Unit) {
    val headers = mapOf(
        "Content-Type" to "application/x-www-form-urlencoded",
        "Authorization" to "Basic ${Base64.getEncoder().encodeToString(stripeKeySecret.toByteArray(Charsets.UTF_8))}"
    )
    val body = mapOf(
        "amount" to amount,
        "currency" to currency,
        "automatic_payment_methods[enabled]" to "true"
    )

    performPostRequest(clientSecretURL, headers, body, completion)
}